yktracec/src/jitmodbuilder.cc

```diff
@@ -1266,6 +1293,8 @@ class JITModBuilder {
               // setjmp/longjmp, so for now simply abort this trace.
               // See: https://github.com/ykjit/yk/issues/610
               return nullptr;
+            } else if (S == "yk_trace_basicblock") {
```

Comment:
I *think* this check (well, "filter") should be in https://github.com/ykjit/yk/pull/939/files#diff-b952f4a017fefda851303dbeff602e0064b56a0da470e4e64de9c35fd950c8e4: there's no need for jitmodbuilder to know about it, as it's an internal swt detail.

Reply:
Makes sense.
I'll remove it from `jitmodbuilder.cc`

Reply:
I don't see how this can be moved to `swt/mod.rs`. All this does is skip the `yk_trace_basicblock` call instruction so it isn't copied over into the JIT module. I don't think `swt/mod.rs` has a view of the AOT IR in order to do the same.

Reply:
I don't know the quickest/easiest way for swt/mod.rs to do this, but logically it should be its job. If we're not careful, jitmodbuilder (or its modern equivalent) will just end up knowing too much about the quirks of each tracing backend. [That said, we might not be able to fully get away this, and I don't want to torture ourselves over it. So we'll have to be pragmatic.]

Reply:
In order to do this `swt/mod.rs` would have to iterate over all instructions and check if they are a call to `yk_trace_basicblock` and then remove them. That being said, I believe that call is always the first instruction of each block (@Pavel-Durov will need to double-check that), in which case you could just remove the first instruction of each block before handing things over to `jitmodbuilder`.

Reply:
> In order to do this swt/mod.rs would have to iterate over all instructions and check if they are a call to yk_trace_basicblock and then remove them.

Yes. So the real question is: is there a fast way to do this? I can't remember how swt identifies blocks. If it does it by address (somehow) it might be as simple as comparing against the symbol address of the `yk_trace_basicblock` function header or similar?

>That being said, I believe that call is always the first instruction of each block (@Pavel-Durov will need to double-check that), in which case you could just remove the first instruction of each block before handing things over to jitmodbuilder.

Or this! Though I think we'd want to leave lots of `debug_assert`s in to check this works OK.

Reply:
Correction: the above isn't correct. If a block has phi-nodes or landing-pad instructions, the `yk_trace_basicblock` call will be inserted afterwards. So it can happen that it is not the first instruction of a block.

Reply:
> Yes. So the real question is: is there a fast way to do this? I can't remember how swt identifies blocks. If it does it by address (somehow) it might be as simple as comparing against the symbol address of the `yk_trace_basicblock` function header or similar?

I don't see how that would help.  I had another look at `swt/mod.rs` and it doesn't even have a view of the instructions or the LLVM IR. It literally just has a vector of block indidces. Those indices are handed to jitmodbuilder which then extracts the LLVM IR blocks and processes them. There is no way that `swt/mod.rs` can alter instructions unless we change both swt and hwt to hand jitmodbuilder actual LLVM IR blocks instead of just the indices.

Reply:
One way to fix this is to intercept the LLVM IR before it gets serialised and remove those calls then. This is doable for the llvm backend, and should be even easier for our new yk backend as we lower every LLVM instruction to our own AOT IR. So removing a call at the point is very simple.

Reply:
Oh, I overlooked how `jitmodbuilder` works 👀

Reply:
> One way to fix this is to intercept the LLVM IR before it gets serialised and remove those calls then. 

That's a good idea and has other benefits too (e.g. less serialised stuff)!

Reply:
If the main concern here is the separation of concerns, i.e. that jitmodbuilder (or its modern equivalent) will end up knowing too much. We can use an abstraction on the tracer implementation while having concrete implementation encapsulated in the tracers.

Something like:
```
// jitmodbuilder
if (is_tracable_function(yk_S) == false) {
	continue;
}

// swt
fn is_tracable_function(func_name CString) -> boolean {
	return func_name != "yk_trace_basicblock";
}

// hwt
fn is_tracable_function(func_name CString)  -> boolean {
	return true;
}
``` 
But intercepting the LLVM IR before it gets serialised sounds cleaner

Reply:
>  But intercepting the LLVM IR before it gets serialised sounds cleaner

Agreed.